### ISSUE TITLE
Graphite is very slow rendering requests when i use asterisk in the query

### ISSUE THREAD

**Issue by wangzaiqqhop** [graphite-project/graphite-web#2183]
hello all:
     when i use sumSeries(api.*.QOSController.*.category_Q19_count) to request the data from graphite，it goes very slow；but if i change it to sumSeries(api.*.QOSController.A.category_Q19_count,api.*.QOSController.B.category_Q19_count,api.*.QOSController.C.category_Q19_count),the request goes very fast . 
the amount of wsp file under the QOSController folder is about  30 thousand，but i only want to use asterisk to match the A、B、C folder ，i don‘t understand why use the asterisk in the query could be so slowly
      could you help me,thank you!

**Comment by wangzaiqqhop** [comment]
![tim 20180106171903](https://user-images.githubusercontent.com/18474372/34638425-dd50e3ca-f305-11e7-893d-ee18716b6bd9.png)

**Comment by wangzaiqqhop** [comment]
![tim 20180106172125](https://user-images.githubusercontent.com/18474372/34638438-2548e5f6-f306-11e7-9228-9c34a4943268.png)

**Comment by deniszh** [comment]
Hi @wangzaiqqhop,
Could you please explain more about your setup? 
Do you use cluster or single server? What's version of Graphite?

**Comment by wangzaiqqhop** [comment]
@deniszh thank you for answer my question so soon,
i use just a single server，one carbon-cache use to collect the metrics from about 10 online-mechine，receive around 35 thousand datapoint per minute，
the graphite-carbon version is 1.0，
when i use exact query or the node of asterisk has less amount of wsp file，the render request goes very fast；but once the query contines asterisk and the  wsp files under the asterisk node  is very large, the query becomes very slow，it just like the asterisk has slow down the speed of query the location of the target file，could you tell me the principle of graphite use to matching documents while the  asterisk exists or not ?
just like the query sumSeries(api.*.QOSController.*.hx_subCategory_Q19_daily_info1_count),the second asterisk is exactly only need to match A、B、C three categories,and the 'hx_subCategory_Q19_daily_info1_count' is just exists in the A/B/C folder ，but under the node of QOSController there exist other 30 thousand wsp file，i thought whether is because of the large amount of wsp file exist under the QOSController node，so use asterisk to match the A/B/C folder become very slowly，but i exactly don’t understand the principle.
as the picture show the 4 query above is very slow，but the 4 below is very fast.
![slowquery](https://user-images.githubusercontent.com/18474372/34640003-e0c2b2e8-f325-11e7-9863-f6e21a1878e1.png)

**Comment by wangzaiqqhop** [comment]
@deniszh thank you for answer my question so soon,
i use just a single server，one carbon-cache use to collect the metrics from about 10 online-mechine，receive around 35 thousand datapoint per minute，
the graphite-carbon version is 1.0，
when i use exact query or the node of asterisk has less amount of wsp file，the render request goes very fast；but once the query contines asterisk and the wsp files under the asterisk node is very large, the query becomes very slow，it just like the asterisk has slow down the speed of query the location of the target file，could you tell me the principle of graphite use to matching documents while the asterisk exists or not ?
just like the query sumSeries(api..QOSController..hx_subCategory_Q19_daily_info1_count),the second asterisk is exactly only need to match A、B、C three categories,and the 'hx_subCategory_Q19_daily_info1_count' is just exists in the A/B/C folder ，but under the node of QOSController there exist other 30 thousand wsp file，i thought whether is because of the large amount of wsp file exist under the QOSController node，so use asterisk to match the A/B/C folder become very slowly，but i exactly don’t understand the principle.

**Comment by wangzaiqqhop** [comment]
just as the picture show the 4 query above is very slow but the 4 below is very fast
 
![slow](https://user-images.githubusercontent.com/18474372/34640073-e71fc0bc-f326-11e7-8f8e-4e90b07b0689.png)

**Comment by DanCech** [comment]
@wangzaiqqhop when you use `*` the finder has to search every folder that matches, so if you have 30k folders but only want `A`, `B`, & `C` it still has to go through each of the 30k to see if they contain a `hx_subCategory_Q19_daily_info1_count` folder.  You can do it in a single query like `sumSeries(api.*.QOSController.{A,B,C}.hx_subCategory_Q19_daily_info1_count)`

**Comment by wangzaiqqhop** [comment]
@DanCech thank you for your answer!
but i still do not quite understand the principle, except the A/B/C node is folder，the other file under the QOSController is all leaf node，so while the finder match the leaf node it will never continue to operate, i can't understand how could the finder matching the need node so slowly?
add i try to use the query  sumSeries(api.*.QOSController.{A,B,C}.hx_subCategory_Q19_daily_info1_count) it seems has no difference with the sumSeries(api.*.QOSController.*.hx_subCategory_Q19_daily_info1_count) , also is not so fast.
whether graphite has a index list of the wsp files and when it's exact query ，graphite will use hash to position the file directly？and while use asterisk it will match the file under the node one by one ？
thank you!

**Comment by DanCech** [comment]
Yes, I just checked the finder code and it isn't smart enough (right now) to see that it can locate all matches for `{A,B,C}` without scanning the entire directory :(  It's that scan that's slow.

Looking at the code it seems like it can be optimized, but in the meantime you're best off with `sumSeries(api.*.QOSController.A.hx_subCategory_Q19_daily_info1_count,api.*.QOSController.B.hx_subCategory_Q19_daily_info1_count,api.*.QOSController.C.hx_subCategory_Q19_daily_info1_count)`

**Comment by wangzaiqqhop** [comment]
@DanCech oh，thank you！

**Comment by wangzaiqqhop** [comment]
@DanCech hello，DanCech：
could you give me a brief description why the exact query is very fast，my python is not so good ，i have just go the check the finder code but haven't checkout the reason；is there a index file tree in the graphite？i have checked the  path '/opt/graphite/storage/ceres',  nothing is in it，or is there some other reason ，why exact query has no need to scan the entire directory and could position the target file so quickly?
thank you!

**Comment by DanCech** [comment]
https://github.com/graphite-project/graphite-web/blob/master/webapp/graphite/finders/standard.py#L97-L170

It's a little hard to follow since it operates recursively, but you're interested in `has_wildcard` on line 103-104, that is what triggers it to either look for a specific file/directory or scan the folder.

I think that it'd be possible to restructure the whole thing to make it more efficient by using the objects returned by `scandir`, and possibly to see whether an expression like `{A,B,C}` has a finite set of matches and if so testing them instead of scanning.

**Comment by wangzaiqqhop** [comment]
@DanCech ok,thank you for your answer!

**Comment by wangzaiqqhop** [comment]
@DanCech hello，DanCech：
      i have changed the  finder part of graphite to your new commit code，and the query  efficiency has improved about 60 percent，that is great；
      according to your last commit ，may i think that you have changed the logic to avoid some leaf node execution the method of join？
![join](https://user-images.githubusercontent.com/18474372/34652088-8f8e1b98-f414-11e7-988d-dfd711449b7f.png)
      i tested the finder code today and found that it seems like the main time cost on the fuzzy query now， is because of the folder scan ，i am thinking is it possible to cache the filename list of each folder in a container such as a hash Set for some time，to avoid the folder scan of the subsequent queries?
      thank you!

**Comment by wangzaiqqhop** [comment]
the join method:
![join](https://user-images.githubusercontent.com/18474372/34652178-b9d1d92a-f415-11e7-9eb3-db816200bd54.png)

**Comment by wangzaiqqhop** [comment]
![join](https://user-images.githubusercontent.com/18474372/34652181-cc6377c4-f415-11e7-87e6-16ffd12d040c.png)

**Comment by wangzaiqqhop** [comment]
![tim 20180108014835](https://user-images.githubusercontent.com/18474372/34652192-14fee702-f416-11e7-8ba3-5e85a39b3ae4.png)
the join method

**Comment by deniszh** [comment]
fixed in master, backported to stable branch, will be released soon.
